### ISSUE TITLE
Bug: etcd restore failure can end in deletion of all kubernetes containers

### ISSUE THREAD

**Issue by patrick0057** [rancher/rancher#23121]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
Attempt to restore etcd snapshot from web interface.

**Result:**
If your cluster is having some issues, the etcd restore could fail and leave all of your nodes without any kubernetes components and no way to restore them.  

**Other details that may be helpful:**
The most recent scenario I've seen this was in a cluster that was having connectivity issues to the Rancher master cluster.  These connectivity issues turned out to be a blockage in AWS preventing traffic between the two VPC's.  Once this was fixed, the etcd restore kicked in and failed.  This left the nodes in the downstream cluster with several agent containers running per a node and 0 kubernetes components (kubelet, kube-apiserver, etcd, kube-proxy, kube-scheduler, kube-controller-manager).  The containers were deleted, not a in a stopped state.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.2.8
- Installation option (single install/HA):HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 2 Core / 8 GB
- Kubernetes version (use `kubectl version`): 1.14.6

- Docker version (use `docker version`):

```
Containers: 25
 Running: 4
 Paused: 0
 Stopped: 21
Images: 26
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: bb71b10fd8f58240ca47fbb579b9d1028eea7c84
runc version: 2b18fe1d885ee5083ef9f0838fee39b62d653e30
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-1050-aws
Operating System: Ubuntu 18.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.789GiB
Name: <redacted>
ID: 5R5B:O4X7:3XOL:CWPI:ZOLZ:UB7S:HSU7:F6MT:XKZT:4DQN:QMUC:MQ5L
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
HTTP Proxy: <redacted>
HTTPS Proxy: <redacted>
No Proxy: <redacted>
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
 
WARNING: No swap limit support
```

**Comment by patrick0057** [comment]
After further troubleshooting it seems that kube-apiserver in this environment was not able to read secrets properly due to utilizing an encryption provider with the following option passed to the kube-apiserver.

--encryption-provider-config=/etc/kubernetes/encryption.yaml

When the restore took place, the /etc/kubernetes/ directory was wiped out which caused the encryption stop working which in turn caused the restore to fail.  We only discovered this after bootstrapping a controlplane,etcd node manually.  It wasn't working and was noted that encryption was supposed to be enabled but it wasn't due to the missing file.  Once replaced that file we were able to finish the manual bootstrap, restore of the node and then rebuild the cluster to a working state using the web interface.

I've seen other cluster issues cause etcd restores to leave the cluster in a containerless state but I'm not sure what the cause was in those past incidents.  For this issue it would be helpful to have some sort of logic to restore the kubernetes containers when the restore fails so that it is easier to recover.

To prevent this specific incident from happening, would it be possible to change the logic to not delete the whole directory but rather only replace files in /etc/kubernetes/ that need to be replaced for the restore?
